sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuDeviceManager.scala

```diff
@@ -500,12 +442,110 @@ object GpuDeviceManager extends Logging {
+    val heapSizeKey = "spark.executor.memory"
+    val sparkOffHeapEnabledKey = "spark.memory.offHeap.enabled"
+    val sparkOffHeapSizeKey = "spark.memory.offHeap.size"
+
+    def toBytes = ConfHelper.byteFromString(_, ByteUnit.BYTE)
+
+    val executorOverhead = sparkConf.getOption(executorOverheadKey).map(toBytes)
+    val pysparkOverhead = toBytes(sparkConf.get(pysparkOverheadKey, "0"))
+    val heapSize = toBytes(sparkConf.get(heapSizeKey, "1g"))
+    val sparkOffHeapEnabled = sparkConf.getBoolean(sparkOffHeapEnabledKey, defaultValue = false)
+    val sparkOffHeapSize = if (sparkOffHeapEnabled) {
+      toBytes(sparkConf.get(sparkOffHeapSizeKey, "0"))
+    } else {
+      0L
+    }
+
+    val memoryLimit = if (confLimit.isDefined) {
+      if (executorOverhead.isEmpty) {
+        logWarning(s"$executorOverheadKey is not set")
+      }
+      logInfo(s"using configured ${RapidsConf.OFF_HEAP_LIMIT_SIZE} of ${confLimit.get}")
+      confLimit.get
+    } else {
+      // in case we cannot query the host for available memory due to environmental
+      // constraints, we can fall back to minMemoryLimit via saying there's no available
+      lazy val availableHostMemory = memCheck.getAvailableMemoryBytes(conf).getOrElse(0L)
+      val hostMemUsageFraction = .8
+      lazy val basedOnHostMemory = (hostMemUsageFraction * (availableHostMemory - heapSize
+        - pysparkOverhead - sparkOffHeapSize) / deviceCount).toLong
+      if (executorOverhead.isDefined) {
+        val basedOnConfiguredOverhead = executorOverhead.get - pysparkOverhead - sparkOffHeapSize
+        logWarning(s"${RapidsConf.OFF_HEAP_LIMIT_SIZE} is not set; we derived " +
+          s"a memory limit from ($executorOverheadKey - " + s"$pysparkOverheadKey -" +
+          s"$sparkOffHeapSizeKey) = (${executorOverhead.get} - $pysparkOverhead - " +
+          s"$sparkOffHeapSize) = $basedOnConfiguredOverhead")
+        if (basedOnConfiguredOverhead < minMemoryLimit) {
+          logWarning(s"memory limit $basedOnConfiguredOverhead is less than the minimum of " +
+            s"$minMemoryLimit; using the latter")
+          if (minMemoryLimit > basedOnHostMemory) {
+            logWarning(s"the amount of available memory detected on the host is " +
+              s"$basedOnHostMemory, which is less than the minimum")
+          }
+          minMemoryLimit
+        } else {
+          basedOnConfiguredOverhead
+        }
+      } else {
+        logWarning(s"${RapidsConf.OFF_HEAP_LIMIT_SIZE} is not set; we used " +
+          s"memory limit derived from .8 * (estimated available host memory - " +
+          s"$heapSizeKey - $pysparkOverheadKey - $sparkOffHeapSizeKey) = " +
+          s".8 * ($availableHostMemory - $heapSize - $pysparkOverhead - $sparkOffHeapSize) " +
+          s"= $basedOnHostMemory")
+        if (basedOnHostMemory < minMemoryLimit) {
+          logWarning(s"the memory limit based on host memory of $basedOnHostMemory " +
```

Comment:
similar comment on this warning, it would be nice to call out the difference between "available" and "computed" host memory.